stan/math/prim/scal/fun/grad_F32.hpp

```diff
@@ -1,57 +1,127 @@
 
-        tNew = exp(logT);
+      T log_t_old = 0.0;
+      T log_t_new = 0.0;
 
-        gOld[0] = tNew * (gOld[0] / tOld + 1.0 / (a + k));
-        gOld[1] = tNew * (gOld[1] / tOld + 1.0 / (b + k));
-        gOld[2] = tNew * (gOld[2] / tOld + 1.0 / (c + k));
+      T log_z = log(z);
 
-        gOld[3] = tNew * (gOld[3] / tOld - 1.0 / (d + k));
-        gOld[4] = tNew * (gOld[4] / tOld - 1.0 / (e + k));
+      T p = 0.0;
 
-        gOld[5] = tNew * (gOld[5] / tOld + 1.0 / z);
-
-        for (int i = 0; i < 6; ++i) g[i] += gOld[i];
-
-        tOld = tNew;
+      double log_t_new_sign = 1.0;
+      double log_t_old_sign = 1.0;
+      double log_g_old_sign[6];
+      for (double *q = log_g_old_sign; q != log_g_old_sign + 6; ++q)
+        *q = 1.0;
 
+      int k = 0;
+      T term;
+      while (true) {
+        p = (a1 + k) * (a2 + k) * (a3 + k) / ((b1 + k) * (b2 + k) * (1 + k));
+        if (p == 0)
+          break;
+
+        log_t_new += log(fabs(p)) + log_z;
+        log_t_new_sign *= sign(p);
+
+//        g_old[0] = t_new * (g_old[0] / t_old + 1.0 / (a1 + k));
+        term = log_g_old_sign[0] * log_t_old_sign *
+          exp(log_g_old[0] - log_t_old) + 1/(a1 + k);
+        log_g_old[0] = log_t_new + log(fabs(term));
+        log_g_old_sign[0] = sign(term) * log_t_new_sign;
+
+//        g_old[1] = t_new * (g_old[1] / t_old + 1.0 / (a2 + k));
+        term = log_g_old_sign[1] * log_t_old_sign *
+          exp(log_g_old[1] - log_t_old) + 1/(a2 + k);
+        log_g_old[1] = log_t_new + log(fabs(term));
+        log_g_old_sign[1] = sign(term) * log_t_new_sign;
+
+//        g_old[2] = t_new * (g_old[2] / t_old + 1.0 / (a3 + k));
+        term = log_g_old_sign[2] * log_t_old_sign *
+          exp(log_g_old[2] - log_t_old) + 1/(a3 + k);
+        log_g_old[2] = log_t_new + log(fabs(term));
+        log_g_old_sign[2] = sign(term) * log_t_new_sign;
+
+//        g_old[3] = t_new * (g_old[3] / t_old - 1.0 / (b1 + k));
+        term = log_g_old_sign[3] * log_t_old_sign *
+          exp(log_g_old[3] - log_t_old) - 1/(b1 + k);
+        log_g_old[3] = log_t_new + log(fabs(term));
+        log_g_old_sign[3] = sign(term) * log_t_new_sign;
+
+//        g_old[4] = t_new * (g_old[4] / t_old - 1.0 / (b2 + k));
+        term = log_g_old_sign[4] * log_t_old_sign *
+          exp(log_g_old[4] - log_t_old) - 1/(b2 + k);
+        log_g_old[4] = log_t_new + log(fabs(term));
+        log_g_old_sign[4] = sign(term) * log_t_new_sign;
+
+//        g_old[5] = t_new * (g_old[5] / t_old + 1.0 / z);
+        term = log_g_old_sign[5] * log_t_old_sign *
+          exp(log_g_old[5] - log_t_old) + 1/z;
+        log_g_old[5] = log_t_new + log(fabs(term));
+        log_g_old_sign[5] = sign(term) * log_t_new_sign;
+
+        for (int i = 0; i < 6; ++i) {
+          g[i] += log_g_old_sign[i] * exp(log_g_old[i]);
+        }
+
+        if (exp(log_t_new) <= precision)
+          break;  // implicit abs
+
+        if (k >= max_steps) {
```

Comment:
this needs to throw as soon as k exceeds the threshold so I can't pull it out of the loop, though it would make sense in the for loop you suggested so maybe that's what you mean? Still sticking to Michael's suggested construct for now.